### PR TITLE
ICU-21526 Change return of UElementComparator to int32_t

### PR DESCRIPTION
**Pull request by FrankYFTang**
##### Checklist

- [X] Required: Issue filed: https://unicode-org.atlassian.net/browse/ICU-21526
- [X] Required: The PR title must be prefixed with a JIRA Issue number. 
- [X] Required: The PR description must include the link to the Jira Issue, for example by completing the URL in the first checklist item
- [X] Required: Each commit message must be prefixed with a JIRA Issue number. 
- [x] Issue accepted (done by Technical Committee after discussion)
- [X] Tests included, if applicable
- [x] API docs and/or User Guide docs changed or added, if applicable

### CODE REVIEW

**Comment by FrankYFTang** [comment]
not sure we should land this for 69.1. maybe we should wait till the tree open of 70.1

**Comment by markusicu** [comment]
> not sure we should land this for 69.1. maybe we should wait till the tree open of 70.1

Yes, this would be an enhancement (making code less brittle), not a bug fix.
And once we get there I will have comments :-)

File: icu4c/source/common/locid.cpp [hunk 1]
```diff
@@ -1576,10 +1576,8 @@ AliasReplacer::replaceTransformedExtensions(
...
         } while (tkey != nullptr);
-        tfields.sort([](UElement e1, UElement e2) -> int8_t {
-            // uprv_strcmp return int and in some platform, such as arm64-v8a,
-            // it may return positive values > 127 which cause the casted value
-            // of int8_t negative.
+        tfields.sort([](UElement e1, UElement e2) -> int32_t {
+            // uprv_strcmp may return value other than {-1, 0, 1}.
             int res = uprv_strcmp(
                 (const char*)e1.pointer, (const char*)e2.pointer);
             return (res == 0) ? 0 : ((res > 0) ? 1 : -1);
```
Comment (markusicu) [inline icu4c/source/common/locid.cpp:1583]:
```suggestion
            return uprv_strcmp((const char*)e1.pointer, (const char*)e2.pointer);
```
I am pretty sure that strcmp() at worst returns the difference between unequal bytes, which means that it needs 9 bits, which easily fit into an int32_t. And usually an int is the same as an int32_t.

**Review by markusicu — COMMENTED** [review]
Now that ICU 69 is off on the maintenance branch, we can proceed here targeting master/main.

Please change the PR title to "ICU-21526 Change return of UElementComparator to int32_t", and once you squash please use this for the commit message, too.

**Comment by FrankYFTang** [comment]
got error on icu4c-release-tools build 
```
Error:  Failed to execute goal on project icu4c-apireport: Could not resolve dependencies for project org.unicode.icu:icu4c-apireport:jar:1.0-SNAPSHOT: Could not transfer artifact xalan:xalan:jar:2.7.2 from/to central (https://repo.maven.apache.org/maven2): Transfer failed for https://repo.maven.apache.org/maven2/xalan/xalan/2.7.2/xalan-2.7.2.jar 502 Gateway Error -> [Help 1]
Error:  
Error:  To see the full stack trace of the errors, re-run Maven with the -e switch.
Error:  Re-run Maven using the -X switch to enable full debug logging.
Error:  
Error:  For more information about the errors and possible solutions, please read the following articles:
Error:  [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/DependencyResolutionException
Error: Process completed with exit code 1.
```
Not sure how to fix this

File: icu4c/source/common/locid.cpp (hunk 1 above)
Comment (FrankYFTang) [reply icu4c/source/common/locid.cpp:1583]: but that does not mean it will return 0, -1 and 1 which is still mandated by UElementComparator

Comment (markusicu) [reply icu4c/source/common/locid.cpp:1583]: Yes, see my [ticket comment from March 5](https://unicode-org.atlassian.net/browse/ICU-21526?focusedCommentId=160438). Please take a look at the UElementComparator call sites, I bet that few or none rely on these three values.

Comment (FrankYFTang) [reply icu4c/source/common/locid.cpp:1583]: I understand but icu4c/source/common/uelement.h still state

```
 * @return 0 if the two elements are equal, -1 if e1 is < e2, or +1 if e1 is > e2.
```

do you want me to change it to
```
 * @return 0 if the two elements are equal, a negative value  if e1 is < e2, or a positive value if e1 is > e2.
```
? If not changing there, then we cannot drop this change here, right?

Comment (markusicu) [reply icu4c/source/common/locid.cpp:1583]: Please look at the call sites. If they don't actually need the return value to be one of these three values, or if they are easy to fix, then please change the "API" docs.

**Review by markusicu — COMMENTED** [review]
lgtm pse squash

**Review by markusicu — APPROVED** [review]
lgtm

**Pull request merged**